lib/lambda_ethereum_consensus/state_transition/predicates.ex

```diff
@@ -102,6 +102,34 @@ defmodule LambdaEthereumConsensus.StateTransition.Predicates do
     is_indices_available(validators, indices, h < validators)
   end
 
+  @doc """
+  Check if merkle branch is valid
+  """
+  @spec is_valid_merkle_branch(
+          SszTypes.bytes32(),
+          list(SszTypes.bytes32()),
+          SszTypes.uint64(),
+          SszTypes.uint64(),
+          SszTypes.root()
+        ) :: boolean
+  def is_valid_merkle_branch(leaf, branch, depth, index, root) do
+    is_valid_merkle_branch(leaf, branch, depth, index, root, leaf, 0)
+  end
+
+  defp is_valid_merkle_branch(leaf, branch, depth, index, root, value, i) do
+    if i == depth do
+      value == root
+    else
+      if rem(div(index, 2 ** i), 2) == 1 do
+        value = :crypto.hash(:sha256, Enum.at(branch, i) <> value)
+        is_valid_merkle_branch(leaf, branch, depth, index, root, value, i + 1)
+      else
+        value = :crypto.hash(:sha256, value <> Enum.at(branch, i))
+        is_valid_merkle_branch(leaf, branch, depth, index, root, value, i + 1)
+      end
+    end
+  end
```

Comment:
Let's change this to:

```elixir
  defp is_valid_merkle_branch(_, _, depth, _, root, value, depth), do: value == root

defp is_valid_merkle_branch(leaf, branch, depth, index, root, value, i) do
    if rem(div(index, 2 ** i), 2) == 1 do
      value = :crypto.hash(:sha256, Enum.at(branch, i) <> value)
      is_valid_merkle_branch(leaf, branch, depth, index, root, value, i + 1)
    else
      value = :crypto.hash(:sha256, value <> Enum.at(branch, i))
      is_valid_merkle_branch(leaf, branch, depth, index, root, value, i + 1)
    end
  end
```

Then we can refactor the `rem(div(index, 2 ** i), 2) == 1` part to be able to use it in guards. For that, we should divide `index` by 2 at each step, to avoid having to call `**`.

Reply:
I tried it https://github.com/lambdaclass/lambda_ethereum_consensus/pull/417#discussion_r1394657081 in here but the thing is that it appears that is not allowed to use rem() or div() in a guard

Reply:
The one that's not allowed on guards is the `**` operator. `div` and `rem` are.